Source/MLX/MLXArray+Indexing.swift

```diff
@@ -521,4 +353,1209 @@ extension MLXArray {
+        }
+    }
+
+}
+
+// MARK: - Support
+
+func countNonNewAxisOperations(_ operations: any Sequence<MLXArrayIndexOperation>) -> Int {
+    operations
+        .filter { !$0.isNewAxis }
+        .count
+}
+
+/// Replace `.ellipsis` operations with `.slice` operations for the expanded axes.
+///
+/// - Parameters:
+///   - shape: array shape
+///   - operations: operations
+/// - Returns: new operations with `.ellipsis` replaced
+func expandEllipsisOperations(shape: [Int32], operations: [MLXArrayIndexOperation])
+    -> [MLXArrayIndexOperation]
+{
+    // equivalent of mlx_expand_ellipsis
+
+    let ellipsisCount = operations.lazy.filter { $0.isEllipsis }.count
+    if ellipsisCount == 0 {
+        return operations
+    }
+    if ellipsisCount > 1 {
+        fatalError("multiple .ellipsis not allowed: \(operations)")
+    }
+
+    let prefix = operations.prefix { !$0.isEllipsis }
+    let suffix = operations.suffix(from: prefix.count + 1)
+
+    // inject full range slices in place of the ellipsis
+    let expandRange =
+        countNonNewAxisOperations(prefix) ..< (shape.count - countNonNewAxisOperations(suffix))
+    let expand = expandRange.map {
+        MLXArrayIndexOperation.slice(.init(start: 0, end: shape[$0], stride: 1))
+    }
+
+    return prefix + expand + suffix
+}
+
+// MARK: - index get
+
+/// Single operation MLXArray subscript get.
+///
+/// This provides optimized operations for simple forms of subscripts.
+///
+/// - Parameters:
+///   - src: input array
+///   - operation: single index operation
+/// - Returns: MLXArray get result
+func getItem(src: MLXArray, operation: MLXArrayIndexOperation, stream: StreamOrDevice = .default)
+    -> MLXArray
+{
+    switch operation {
+    case .ellipsis:
+        return src
+
+    case .newAxis:
+        return src.expandedDimensions(axis: 0, stream: stream)
+
+    case .index(let index):
+        return src.take(src.resolve(index: index, axis: 0), axis: 0, stream: stream)
+
+    case .slice(let slice):
+        let ndim = src.ndim
+        var starts = [Int32](repeating: 0, count: ndim)
+        var ends = src.shape.asInt32
+        var strides = [Int32](repeating: 1, count: ndim)
+
+        let size = ends[0]
+        starts[0] = slice.start(size)
+        ends[0] = slice.end(size)
+        strides[0] = slice.stride
+
+        return MLXArray(
+            mlx_slice(
+                src.ctx, starts, starts.count, ends, ends.count, strides, strides.count,
+                stream.ctx))
+
+    case .array(let indices):
+        return src.take(indices, axis: 0, stream: stream)
+    }
+}
+
+/// Full implementation of python/numpy indexing.
+///
+/// - Parameters:
+///   - array: source array
+///   - operations: array of index operations
+/// - Returns: array with the indices applied
+func getItemND(
+    src: MLXArray, operations: [MLXArrayIndexOperation], stream: StreamOrDevice = .default
+) -> MLXArray {
```

Comment:
All of this code is roughly a line-by-line port of the indexing c++ code.  Some variables were renamed for swift and many array operations were rewritten to do the same thing (c++ vector and swift array code is very different!).  The comments are copied across (and edited as needed).  Some additional comments were added as I needed.